sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/trace/samplers/CompositeSampler.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.sdk.extension.incubator.trace.samplers;
+
+import static io.opentelemetry.sdk.extension.incubator.trace.samplers.ImmutableSamplingIntent.INVALID_THRESHOLD;
+import static io.opentelemetry.sdk.extension.incubator.trace.samplers.ImmutableSamplingIntent.isValidRandomValue;
+import static io.opentelemetry.sdk.extension.incubator.trace.samplers.ImmutableSamplingIntent.isValidThreshold;
+import static io.opentelemetry.sdk.extension.incubator.trace.samplers.OtelTraceState.OTEL_TRACE_STATE_KEY;
+
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.internal.OtelEncodingUtils;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.SpanKind;
+import io.opentelemetry.api.trace.TraceState;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.sdk.trace.data.LinkData;
+import io.opentelemetry.sdk.trace.samplers.Sampler;
+import io.opentelemetry.sdk.trace.samplers.SamplingDecision;
+import io.opentelemetry.sdk.trace.samplers.SamplingResult;
+import java.util.List;
+
+/**
+ * A sampler that uses a {@link ComposableSampler} to make its sampling decisions while handlign
+ * tracestate.
+ */
+public final class CompositeSampler implements Sampler {
+  /**
+   * Returns a new composite {@link Sampler} that delegates to the given {@link ComposableSampler}.
+   */
+  public static Sampler wrap(ComposableSampler delegate) {
+    return new CompositeSampler(delegate);
+  }
+
+  private final ComposableSampler delegate;
+
+  private CompositeSampler(ComposableSampler delegate) {
+    this.delegate = delegate;
+  }
+
+  @Override
+  public SamplingResult shouldSample(
+      Context parentContext,
+      String traceId,
+      String name,
+      SpanKind spanKind,
+      Attributes attributes,
+      List<LinkData> parentLinks) {
+    TraceState traceState = Span.fromContext(parentContext).getSpanContext().getTraceState();
+    OtelTraceState otTraceState = OtelTraceState.parse(traceState);
+
+    SamplingIntent intent =
+        delegate.getSamplingIntent(parentContext, traceId, name, spanKind, attributes, parentLinks);
+
+    boolean thresholdReliable = false;
+    boolean sampled = false;
+    if (isValidThreshold(intent.getThreshold())) {
+      thresholdReliable = intent.isThresholdReliable();
+      long randomValue;
```

Comment:
If the threshold is not reliable, the sampler must not use the random value from the trace state or from the trace-id. This is because we cannot guarantee that these "random" values have the required uniform distribution over the (MIN_THRESHOLD, MAX_THRESHOLD) interval.

Reply:
I can't find anything in the spec on this. I noticed the caveat to deal with rate limiting sampler but since that's also not in the spec went with the simpler implementation. Is this something to change now or can we follow up after spec clarification?